File: app/src/main/java/com/bitcoin/merchant/app/application/PaymentProcessor.java
package com.bitcoin.merchant.app.application;

import android.content.ContentValues;
import android.util.Log;

import com.bitcoin.merchant.app.database.DBControllerV3;
import com.bitcoin.merchant.app.database.PaymentRecord;
import com.bitcoin.merchant.app.util.AppUtil;
import com.crashlytics.android.Crashlytics;

import org.bitcoindotcom.bchprocessor.bip70.model.InvoiceStatus;
import org.bitcoindotcom.bchprocessor.bip70.model.InvoiceStatusOutput;

import static com.bitcoin.merchant.app.MainActivity.TAG;

public class PaymentProcessor {
    private final CashRegisterApplication app;
    private final DBControllerV3 db;

    public PaymentProcessor(CashRegisterApplication app) {
        this.app = app;
        this.db = app.getDb();
    }

    public ContentValues recordInDatabase(InvoiceStatus i, String fiatFormatted) {
        long bch = i.getTotalBchAmount();
        try {
            if (AppUtil.getPaymentTarget(app).isXPub()) {
                for (InvoiceStatusOutput output : i.outputs) {
                    app.getWallet().addUsedAddress(output.address);
                }
            }
            String message = "";
            int confirmations = 0;
            String addr = i.getFirstAddress();
            long time = System.currentTimeMillis();
            PaymentRecord r = new PaymentRecord(time, addr, bch, fiatFormatted, confirmations, message, i.txId);
            ContentValues values = r.toContentValues();
            db.insertPayment(values);
            return values;
        } catch (Exception e) {
            Log.e(TAG, "recordInDatabase" + i, e);
            Crashlytics.logException(e);
            return null;
        }
    }
}

File: app/src/main/java/com/bitcoin/merchant/app/application/PaymentProcessor.kt
package com.bitcoin.merchant.app.application

import android.util.Log
import com.bitcoin.merchant.app.MainActivity
import com.bitcoin.merchant.app.database.DBControllerV3
import com.bitcoin.merchant.app.database.PaymentRecord
import com.bitcoin.merchant.app.model.Analytics
import com.bitcoin.merchant.app.network.PaymentReceived
import com.bitcoin.merchant.app.screens.dialogs.DialogHelper
import com.bitcoin.merchant.app.util.Settings
import org.bitcoindotcom.bchprocessor.bip70.model.InvoiceStatus

class PaymentProcessor(private val app: CashRegisterApplication, private val db: DBControllerV3) {
    fun recordInDatabase(i: InvoiceStatus, fiatFormatted: String?) {
        val bch = i.totalAmountInSatoshi
        try {
            if (Settings.getPaymentTarget(app).isXPub) {
                for ((_, _, address) in i.outputs) {
                    app.wallet.addUsedAddress(address)
                }
            }
            val message = ""
            val confirmations = 0
            val addr = i.firstAddress
            val timeInSec = System.currentTimeMillis() / 1000
            db.insertPayment(PaymentRecord(timeInSec, addr, bch, fiatFormatted, confirmations, message, i.txId))
        } catch (e: Exception) {
            Analytics.error_db_write_tx.sendError(e)
            Log.e(MainActivity.TAG, "recordInDatabase $i", e)
        }
    }

    fun recordInDatabase(p: PaymentReceived, fiatFormatted: String?) {
        val bch = p.bchReceived
        try {
            if(p.isUnderpayment || p.isOverpayment) {
                if (Settings.getPaymentTarget(app).isXPub) {
                    app.wallet.addUsedAddress(p.addr)
                }
                val message = ""
                val confirmations = -1
                val addr = p.addr
                val timeInSec = System.currentTimeMillis() / 1000
                db.insertPayment(PaymentRecord(timeInSec, addr, bch, fiatFormatted, confirmations, message, p.txHash))
            } else {
                if (Settings.getPaymentTarget(app).isXPub) {
                    app.wallet.addUsedAddress(p.addr)
                }
                val message = ""
                val confirmations = 0
                val addr = p.addr
                val timeInSec = System.currentTimeMillis() / 1000
                db.insertPayment(PaymentRecord(timeInSec, addr, bch, fiatFormatted, confirmations, message, p.txHash))
            }
        } catch (e: Exception) {
            Analytics.error_db_write_tx.sendError(e)
            Log.e(MainActivity.TAG, "recordInDatabase $p", e)
        }
    }

    fun paymentAlreadyRecorded(tx: String): Boolean {
        return db.paymentAlreadyRecorded(tx)
    }
}